### PR TITLE
Multi-threaded Executor starvation fix

### PR DESCRIPTION
**Pull request by HarunTeper**
Pull request addressing the issues in #2360 #2645.

So far, I have added a test that detects starvation in the multi-threaded executor.
This test includes a mutually-exclusive callback group with two timers.
The executor should be alternating between these two tasks, never executing one task twice before the other.

To fix starvation, I have identified the following steps (which I was not able to completely implement yet):

1. Introduce a new mutex (I refer to this as 'notify_mutex_') to the executor.hpp file that is used to guard callback group flags.
2. Introduce a function that locks the 'notify_mutex_', triggers the 'interrupt_guard_condition_' and unlocks the callback group flag (and the 'notify_mutex_' afterwards). Currently, the 'MultiThreadedExecutor::run' function includes the guard condition trigger method, while the 'Executor::execute_any_executable' includes the call that unlocks the callback group. These need to be combined and guarded by the 'notify_mutex_'.
3. The 'Executor::wait_for_work' function needs to be split into two funtions. One may be called 'Executor::prepare_wait_set', which does everything up to (but excluding) the 'wait' function which is currently in the 'wait_for_work' function. The rest of the wait for work function can be kept in the current function. This change is necessary to lock the 'notify_mutex' while a callback is being extracted from the wait set by 'get_next_executable', ensuring that no other thread can change a callback group flag at the same time.
4. The most complex change: The function that collects and adds the entities to the 'wait_set_' needs to be updated. First, the 'wait_result_' should not be reset. Then a function needs to be added, which adds all callback instances from the previous 'wait_result_' to the current 'wait_result_', if they are blocked and not invalid. This new function should be executed after the 'wait_set_.wait' function. Otherwise, the previously blocked jobs would immediately unblock the wait function, as they are already ready when it is called, which would then lead to busy waiting. Furthermore, for this change, the position at which the previous job instances are added also needs to be decided, which would then decide the priority of the jobs after inserting. For this change, it may also be necessary to introduce a variable called 'previous_wait_result_'.

These steps are based on the work I published here:

https://ieeexplore.ieee.org/document/9622336

https://daes.cs.tu-dortmund.de/storages/daes-cs/r/publications/teper2024emsoft_preprint.pdf

I have already tried to implement some of these steps, and I will also commit some of the changes to this fork this week. However, for step 4, I may require some help. I also noticed that my changes break some of the tests are currently part of rclcpp, as I move the functions that set callback group flags and trigger guard conditions.

### CODE REVIEW

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp [hunk 1]
```diff
@@ -102,3 +102,66 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
   executor.add_node(node);
   executor.spin();
 }
+
+/*
+   Test that no tasks are starved
+ */
+TEST_F(TestMultiThreadedExecutor, starvation) {
```
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:109]: why this should be assertion for the test? if we create the custom callback group with `rclcpp::CallbackGroupType::Reentrant` and add it to `create_wall_timer`, this test should pass because there are 2 threads that executor can assign the executables concurrently.
in other word, this is expected behavior that user specifies with `MutuallyExclusive` (default) as we discussed on https://github.com/ros2/rclcpp/issues/2645?

`SingleThreadedExecutor` has the exact same situation like this. if you are trying to fix this timer starvation in the `Executor`, i do not think that is the problem only for `MultiThreadedExecutor`, if the timer callback overruns.

> The executor should be alternating between these two tasks, never executing one task twice before the other.

IMO, i am not sure if this assumption is correct by system. sounds like user requirement, user would want to have a high priority timer which they want it to be executed as fast as possible. in that case, this fix becomes the problem for that requirement. (in that case, what's missing here is priority order user interface?)

CC: @jmachowinski @mjcarroll @alsora

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp [hunk 2]
```diff
@@ -102,3 +102,66 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
...
+  rclcpp::TimerBase::SharedPtr timer_one;
+  rclcpp::TimerBase::SharedPtr timer_two;
+
+  auto timer_one_callback = [&timer_one_count, &timer_two_count]() {
+    std::cout << "Timer one callback executed. Count: "
+              << timer_one_count.load() << std::endl;
+
+    auto start_time = std::chrono::steady_clock::now();
+    while (std::chrono::steady_clock::now() - start_time < 100ms) {
+    }
```
Comment (jmachowinski) [inline rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp]:
```suggestion
   std::this_thread::wait_for(100ms);
```

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp [hunk 3]
```diff
@@ -102,3 +102,66 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
...
+  };
+
+  auto timer_two_callback = [&timer_one_count, &timer_two_count]() {
+    std::cout << "Timer two callback executed. Count: "
+              << timer_two_count.load() << std::endl;
+
+    auto start_time = std::chrono::steady_clock::now();
+    while (std::chrono::steady_clock::now() - start_time < 100ms) {
+    }
```
Comment (jmachowinski) [inline rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp]:
```suggestion
    std::this_thread::wait_for(100ms);
```

**Review by jmachowinski — CHANGES_REQUESTED** [review]

**Comment by jmachowinski** [comment]
The proposed changes, as well as the paper, don't make sense to me. But this might be a wording vs implementation issue. Lets wait for the actual implementation...
@HarunTeper you might want to attend to the next client workgroup meeting, to discuss this in person.

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp (hunk 1 above)
Comment (alsora) [reply rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:109]: In the unit-test you are creating two timers with a period of `0ms`.
In this scenario there's no guarantee that both timers will be called, because they are both always ready all the time. Note that the problem is not that they have the same period, but rather that the period is `0ms`.
Since they are both always ready, only the "first one" will be invoked.

Can you please check what happens if the period is set to `10ms` for both?

**Comment by alsora** [comment]
Hi @HarunTeper,
I agree that this topic seems complex enough that it would be better to have an in-person conversation.

The next Client Library WG meeting will happen Friday 12/20/2024 at 8AM PST.
See here some details, and I will post a reminder next week https://discourse.ros.org/t/next-client-library-wg-meeting-friday-6th-december-2024-8am-pt/40954

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp [hunk 4]
```diff
@@ -102,3 +102,66 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
...
+      rclcpp::shutdown();
+      ASSERT_LE(diff, 1);
+    }
+  };
+
+  timer_one = node->create_wall_timer(0ms, timer_one_callback);
+  timer_two = node->create_wall_timer(0ms, timer_two_callback);
+
+  executor.add_node(node);
+  executor.spin();
```
Comment (alsora) [inline rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:154]: @HarunTeper it would be nice to convert this into a "real" unit-test.

Right now this code will run forever until an assertion fails, so it's not acceptable.
Moreover, we should remove the debug prints.
You should use GTEST assertions and expectations to provide the necessary information (note that you can add log to those; the logs will be printed only when the test fails).

**Comment by HarunTeper** [comment]
> Hi @HarunTeper, I agree that this topic seems complex enough that it would be better to have an in-person conversation.
> 
> The next Client Library WG meeting will happen Friday 12/20/2024 at 8AM PST. See here some details, and I will post a reminder next week https://discourse.ros.org/t/next-client-library-wg-meeting-friday-6th-december-2024-8am-pt/40954

I would likely join the next meeting, and work on the pull request until then. Sorry for not making it to the last one.

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp (hunk 4 above)
Comment (HarunTeper) [reply rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:154]: I have removed the outputs and simply put an assertion. If you want any other outputs, let me know. The test should automatically stop after roughly one second.

File: rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp (hunk 1 above)
Comment (HarunTeper) [reply rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp:109]: I have set the period of the timers to 10ms and the same behavior as 0ms occurs. This is because the execution time of both timers is greater than the periods (and the time it takes for the executor to update the wait set).

**Comment by HarunTeper** [comment]
I have now uploaded my fix for humble to a custom repository:
It provides a devcontainer and VSCode tasks to build and run the test that shows starvation. (After switching branches, you should delete the build log and install folders before rebuilding)

https://github.com/HarunTeper/rclcpp_humble_multithreaded_executor

You can see the changes in the branch called "fix".

I am currently working on applying the same changes to the current version of the multi-threaded executor. However, since Humble, the code of the executor itself, specifically the wait set and wait_for_work functions are very different.

I will try to implement the version that I propose, however, it will be different from my previous solution. For example, I think the current version has threads busy waiting for tasks to become ready if all tasks are blocked but not all threads are currently executing tasks.

If there is another meeting, I will join that one to talk about the executor. In the meantime, I will work on the fix.

**Comment by jmachowinski** [comment]
I had a look at the 'fix' implementation. As far as I can see, this comes down to:
- If there are unprocessed entities of a callback group, don't add any entities of the callback group to the waitset on repoll
- Make unprocessed entities from the last poll 'persistent' / don't clear unprocessed entities of callback groups in execution on repoll
- Use some mutexes to avoid races between unmarking of callback group execution and unprocessed entities updates

While I think, this fix works, the implementation is awkward, in the sense that the code flow is extremely hard to follow. To be fair, the executor code flow isn't great in the first place...
I would also say, that the implementations suffers from unneeded interruptions of the waitset polling and context switches. Also the be fair here, the humble and current implementation both suffer from the same problem.

I would suggest a different approach to the problem:
Create a deque of ready events per callback group.
Have two lists of callback groups
- Idle
- Processing

Poll for events:
Only poll for callback groups from the idle list
After each poll:
- add the events to the corresponding deque
- Move callback groups with ready events into the processing list

Worker threads:
- if processing is empty
  - poll / block if someone else is already polling
  - Wakup workers corresponding to the size of the processing list
  - continue
- remove first entry of processing
- process and remove the first ready entity of the removed callback group
- check if callback group has further pending events
  - if not, add to idle, wakeup poll thread
  - if there are pending events, readd to processing 

This is basically the logic I use in the EventsCBGExecutor.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/multi-threaded-executor-task-starvation-fix/42190/1

**Comment by HarunTeper** [comment]
Hello,

I have added some changes to the structure of the executor.cpp. Mainly I have split the wait_for_work function into prepare_work and wait_for_work_simple (placeholder until wait_for_work can be fully replaced).

I am now working on the following changes:

1. In the prepare_work function, I need to determine which callbacks are still in the wait_result_ data structure. These callbacks are blocked, as their callback group flag should have can_be_taken_from set to false. There needs to be a data structure which stores them to be added to the wait_result later. 

2. After the wait_result_.emplace(wait_set_.wait()) function, the previously saved callbacks need to be added back to the wait_result_

If anyone has some suggestions on how I should access wait_result_ and which data structures I should store, please let me know.
